pootle/apps/pootle_statistics/views.py

```diff
@@ -7,79 +7,101 @@
                                      TOP_CONTRIBUTORS_CHUNK_SIZE)
 from pootle.core.url_helpers import split_pootle_path
-from pootle.core.views import PootleJSONMixin, set_permissions, requires_permission
+from pootle.core.views import PootleJSONMixin
+from pootle_language.views import LanguageBrowseView
+from pootle_project.views import ProjectBrowseView, ProjectsBrowseView
+from pootle_translationproject.views import TPBrowseView, TPBrowseStoreView
 
 from .forms import StatsForm
 
 
-class PootleStatsJSON(PootleJSONMixin, View):
-    form_class = StatsForm
-    content_type = None
-
-    @never_cache
-    @method_decorator(ajax_required)
-    @set_permissions
-    @requires_permission("view")
-    def dispatch(self, request, *args, **kwargs):
-        return super(PootleStatsJSON, self).dispatch(request, *args, **kwargs)
-
-    def get_form(self):
-        return self.form_class(self.request.GET)
-
-    @lru_cache()
-    def get_view(self):
-        stats_form = self.get_form()
-        if not stats_form.is_valid():
-            raise Http404(
-                ValidationError(stats_form.errors).messages)
-
-        self.offset = stats_form.cleaned_data['offset'] or 0
-        self.pootle_path = stats_form.cleaned_data['path']
-        rm = resolve(self.pootle_path)
-        try:
-            view_class = get_callable(rm._func_path)
-            view = view_class(request=self.request, kwargs=rm.kwargs)
-        except ViewDoesNotExist:
-            raise Http404()
-
-        return view
-
-    @property
-    def permission_context(self):
-        return self.get_view().permission_context
-
-    def get(self, *args, **kwargs):
-        context = self.get_context_data()
-        return self.render_to_response(context)
-
-
-class TopContributorsJSON(PootleStatsJSON):
+class ContributorsJSONMixin(PootleJSONMixin):
     def get_context_data(self, *args, **kwargs):
         User = get_user_model()
 
         language_code, project_code = split_pootle_path(self.pootle_path)[:2]
+        offset = self.kwargs.get('offset', 0)
 
         top_scorers = User.top_scorers(
             project=project_code,
             language=language_code,
             limit=TOP_CONTRIBUTORS_CHUNK_SIZE + 1,
-            offset=self.offset,
+            offset=offset,
         )
 
         return get_top_scorers_data(
             top_scorers,
             TOP_CONTRIBUTORS_CHUNK_SIZE
         )
+
+
+class TPContributorsJSON(ContributorsJSONMixin, TPBrowseView):
+    pass
+
+
+class TPStoreContributorsJSON(ContributorsJSONMixin, TPBrowseStoreView):
+    pass
+
+
+class LanguageContributorsJSON(ContributorsJSONMixin, LanguageBrowseView):
+    pass
+
+
+class ProjectContributorsJSON(ContributorsJSONMixin, ProjectBrowseView):
+    pass
+
+
+class ProjectsContributorsJSON(ContributorsJSONMixin, ProjectsBrowseView):
+    pass
+
+
+class TopContributorsJSON(View):
+    form_class = StatsForm
+    content_type = None
+
+    @never_cache
+    @method_decorator(ajax_required)
```

Comment:
k - i understand that this delegates permission checking - but i think in this case we should rename this method to `dispatch` as that is more the place to delegate to other views